domain/schema/model.go

```diff
@@ -393,3 +409,123 @@ CREATE INDEX idx_block_device_machine
 ON block_device_machine (machine_uuid);
 `)
 }
+
+func storageSchema() schema.Patch {
+	return schema.MakePatch(`
+CREATE TABLE storage_pool (
+    uuid TEXT PRIMARY KEY,
+    name TEXT NOT NULL,
+    -- Types are provider sourced, so we do not use a lookup with ID.
+    -- This constitutes "repeating data" and would tend to indicate 
+    -- bad relational design. However we choose that here over the
+    -- burden of:
+    --   - Knowing every possible type up front to populate a look-up or;
+    --   - Sourcing the lookup from the provider and keeping it updated. 
+    type TEXT NOT NULL
+);
+
+CREATE TABLE storage_pool_attribute (
+    storage_pool_uuid TEXT NOT NULL,
+    key               TEXT,
+    value             TEXT,
+    CONSTRAINT       fk_storage_pool_attribute_pool
+        FOREIGN KEY  (storage_pool_uuid)
+        REFERENCES   storage_pool(uuid)
+);
+
+CREATE UNIQUE INDEX idx_storage_pool_attribute
+ON storage_pool_attribute (storage_pool_uuid, key);
+
+CREATE TABLE storage_kind (
+    id   INT PRIMARY KEY,
+    kind TEXT
+);
+
+CREATE UNIQUE INDEX idx_storage_kind
+ON storage_kind (kind);
+
+INSERT INTO storage_kind VALUES
+    (0, 'block'), 
+    (1, 'filesystem');
+
+CREATE TABLE storage_instance (
+    uuid            TEXT PRIMARY KEY,
+    storage_kind_id INT NOT NULL,
+    name            TEXT NOT NULL,
+    life_id         INT NOT NULL,
+    CONSTRAINT       fk_storage_instance_kind
+        FOREIGN KEY  (storage_kind_id)
+        REFERENCES   storage_kind(id),
+    CONSTRAINT       fk_storage_instance_life
+        FOREIGN KEY  (life_id)
+        REFERENCES   life(id)
+);
+
+CREATE TABLE storage_instance_pool (
+    storage_instance_uuid TEXT PRIMARY KEY,
+    storage_pool_uuid     TEXT NOT NULL,
+    CONSTRAINT       fk_storage_instance_pool_instance
+        FOREIGN KEY  (storage_instance_uuid)
+        REFERENCES   storage_instance(uuid),
+    CONSTRAINT       fk_storage_instance_pool_pool
+        FOREIGN KEY  (storage_pool_uuid)
+        REFERENCES   storage_pool(uuid)
+);
+
+-- storage_unit_owner is used to indicate when
+-- a unit is the owner of a storage instance.
+-- This is different to a storage attachment.
+CREATE TABLE storage_unit_owner (
+    storage_instance_uuid TEXT PRIMARY KEY,
+    unit_uuid             TEXT NOT NULL,
+    CONSTRAINT       fk_storage_owner_storage
+        FOREIGN KEY  (storage_instance_uuid)
+        REFERENCES   storage_instance(uuid),
+    CONSTRAINT       fk_storage_owner_unit
+        FOREIGN KEY  (unit_uuid)
+        REFERENCES   unit(uuid)
+);
+
+-- Should owner just be a boolean on this table,
+-- or can a unit be an owner without being attached?
```

Comment:
Remove or answer, this is essentially a TODO.

Reply:
I missed asking @wallyworld about this one.

Reply:
I ended up just removing the comment. Having the join table for owner with a PK on the attachment ID is actually a nice way of ensuring that an instance has only one owner.